File: src/udf/python/gcd.py
from risingwave.udf import udf, UdfServer, ScalarFunction
import pyarrow as pa


@udf(input_types=[pa.int32(), pa.int32()], result_type=pa.int32())
def gcd(x: int, y: int) -> int:
    while y != 0:
        (x, y) = (y, x % y)
    return x


class GCD3(ScalarFunction):
    _name = 'gcd'
    _input_types = [pa.int32(), pa.int32(), pa.int32()]
    _result_type = pa.int32()

    def eval(self, x: int, y: int, z: int) -> int:
        return gcd(gcd(x, y), z)


if __name__ == '__main__':
    server = UdfServer()
    server.add_function(gcd)
    server.add_function(GCD3())
    server.serve()

File: src/udf/java/src/main/java/com/risingwave/functions/example/Gcd.java
package com.risingwave.functions.example;

import com.risingwave.functions.ScalarFunction;

public class Gcd extends ScalarFunction {
    public static int eval(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    // TODO: support multiple eval functions
    // public static int eval(int a, int b, int c) {
    // return eval(eval(a, b), c);
    // }
}